spec/models/users/allowed_to_spec.rb

```diff
@@ -33,642 +33,538 @@
 
-    context 'w/ the user being a member in the project
-             w/o the role having the necessary permission
-             w/ inquiring for a permission that is public' do
-      let(:permission) { :view_project }
+            final_setup_step
+          end
 
-      before do
-        project.public = false
+          it do
+            expect(user).to be_allowed_to(permission, project)
+          end
+        end
 
-        member.save!
+        context 'and requesting a public permission' do
+          let(:permission) { :view_project } # a permission defined as public
 
-        final_setup_step
-      end
+          before do
+            final_setup_step
+          end
 
-      it 'is true' do
-        expect(user).to be_allowed_to(permission, project)
+          it { expect(user).to be_allowed_to(permission, project) }
+        end
       end
     end
 
-    context 'w/o the user being member in the project
-             w/ non member being allowed the action
-             w/ the project being private' do
-      before do
-        project.public = false
-        project.save!
+    context 'with the user being a member in the project' do
+      before { member.save! }
 
-        non_member = Role.non_member
+      context 'without the role granting the requested permission' do
+        before do
+          role.remove_permission!(permission)
+        end
 
-        non_member.add_permission! permission
+        context 'and no permissions being assigned to the non-member role' do
+          before { final_setup_step }
 
-        final_setup_step
-      end
+          it { expect(user).not_to be_allowed_to(permission, project) }
+        end
 
-      it 'is false' do
-        expect(user).not_to be_allowed_to(permission, project)
-      end
-    end
+        context 'and the permission being assigend to the non-member role' do
+          before do
+            non_member = Role.non_member
+            non_member.add_permission! permission
 
-    context 'w/o the user being member in the project
-             w/ the project being public
-             w/ non members being allowed the action' do
-      before do
-        project.public = true
-        project.save!
+            final_setup_step
+          end
 
-        non_member = Role.non_member
+          it do
+            # TODO: Figure this one out. In the documentation it says:
+            #
+            # Surprisingly, when looking up permissions, the non member role is always factored in, even it the user
+            # does have other roles within the project as well. That means that if the user has a role in a project
+            # not granting "Create new work package", but the non member role is granting the permission, the user
+            # will in effect have that permission.
+            #
+            # So, in my eyes, with the permission being granted to the non-member role, it should be allowed to do the
+            # requested action.
```

Comment:
I believe the documentation is wrong/outdated. I was also of the impression that the non member role would be factored in up until recently but when trying it and digging into the code I came to the conclusion that it in fact is not taken into account.

Reply:
Removed the spec and removed the sentence from the docs